File: android/app/src/main/java/com/signal/app/features/signal/data/services/EnhancedGeographicService.kt
package com.signal.app.features.signal.data.services

import com.signal.app.core.network.ApiClient
import com.signal.app.core.network.ApiException
import com.signal.app.features.signal.data.models.Signal
import com.signal.app.features.signal.data.models.SignalWithDistance
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import javax.inject.Inject
import javax.inject.Singleton

/**
 * PostGIS 기반 고급 지리적 검색 서비스
 */
@Singleton
class EnhancedGeographicService @Inject constructor(
    private val apiClient: ApiClient
) {

    /**
     * 고급 지리적 검색
     */
    suspend fun advancedSearch(
        latitude: Double,
        longitude: Double,
        radius: Double? = null,
        minRadius: Double? = null,
        maxRadius: Double? = null,
        categories: List<String>? = null,
        searchType: String = "radius",
        polygon: List<List<Double>>? = null,
        route: RouteSearchParams? = null,
        timeFilter: TimeFilterParams? = null,
        density: Boolean = false,
        clustering: ClusteringParams? = null,
        page: Int = 1,
        limit: Int = 20
    ): GeographicSearchResult = withContext(Dispatchers.IO) {
        try {
            val requestBody = buildMap<String, Any> {
                put("latitude", latitude)
                put("longitude", longitude)
                put("radius", radius ?: 5000.0)
                minRadius?.let { put("min_radius", it) }
                maxRadius?.let { put("max_radius", it) }
                categories?.takeIf { it.isNotEmpty() }?.let { put("categories", it) }
                put("search_type", searchType)
                polygon?.let { put("polygon", it) }
                route?.let { put("route", it.toMap()) }
                timeFilter?.let { put("time_filter", it.toMap()) }
                put("density", density)
                clustering?.let { put("clustering", it.toMap()) }
                put("page", page)
                put("limit", limit)
            }

            val response = apiClient.post<Map<String, Any>>(
                "/geographic/search",
                requestBody
            )

            GeographicSearchResult.fromMap(response)
        } catch (e: Exception) {
            throw ApiException("고급 검색에 실패했습니다: ${e.message}")
        }
    }

    /**
     * 밀도 맵 조회
     */
    suspend fun getDensityMap(
        latitude: Double,
        longitude: Double,
        radius: Double = 5000.0,
        gridSize: Double = 0.01
    ): List<DensityPoint> = withContext(Dispatchers.IO) {
        try {
            val queryParams = mapOf(
                "lat" to latitude.toString(),
                "lon" to longitude.toString(),
                "radius" to radius.toString(),
                "grid_size" to gridSize.toString()
            )

            val response = apiClient.get<Map<String, Any>>(
                "/geographic/density",
                queryParams
            )

            @Suppress("UNCHECKED_CAST")
            val densityData = response["data"] as List<Map<String, Any>>
            densityData.map { DensityPoint.fromMap(it) }
        } catch (e: Exception) {
            throw ApiException("밀도 맵 조회에 실패했습니다: ${e.message}")
        }
    }

    /**
     * 핫스팟 분석
     */
    suspend fun getHotspots(
        latitude: Double,
        longitude: Double,
        radius: Double = 10000.0,
        minSignals: Int = 5
    ): List<Hotspot> = withContext(Dispatchers.IO) {
        try {
            val queryParams = mapOf(
                "lat" to latitude.toString(),
                "lon" to longitude.toString(),
                "radius" to radius.toString(),
                "min_signals" to minSignals.toString()
            )

            val response = apiClient.get<Map<String, Any>>(
                "/geographic/hotspots",
                queryParams
            )

            @Suppress("UNCHECKED_CAST")
            val hotspotsData = response["data"] as List<Map<String, Any>>
            hotspotsData.map { Hotspot.fromMap(it) }
        } catch (e: Exception) {
            throw ApiException("핫스팟 분석에 실패했습니다: ${e.message}")
        }
    }

    /**
     * 다각형 영역 검색
     */
    suspend fun searchInPolygon(
        polygon: List<List<Double>>,
        categories: List<String>? = null,
        page: Int = 1,
        limit: Int = 20
    ): List<SignalWithDistance> = withContext(Dispatchers.IO) {
        try {
            val requestBody = buildMap<String, Any> {
                put("polygon", polygon)
                categories?.let { put("categories", it) }
                put("page", page)
                put("limit", limit)
            }

            val response = apiClient.post<Map<String, Any>>(
                "/geographic/polygon",
                requestBody
            )

            @Suppress("UNCHECKED_CAST")
            val signalsData = response["data"] as List<Map<String, Any>>
            signalsData.map { SignalWithDistance.fromMap(it) }
        } catch (e: Exception) {
            throw ApiException("다각형 검색에 실패했습니다: ${e.message}")
        }
    }

    /**
     * 경로 기반 검색
     */
    suspend fun searchAlongRoute(
        startLat: Double,
        startLon: Double,
        endLat: Double,
        endLon: Double,
        bufferWidth: Double = 1000.0,
        categories: List<String>? = null,
        page: Int = 1,
        limit: Int = 20
    ): List<SignalWithDistance> = withContext(Dispatchers.IO) {
        try {
            val requestBody = buildMap<String, Any> {
                put("start_lat", startLat)
                put("start_lon", startLon)
                put("end_lat", endLat)
                put("end_lon", endLon)
                put("buffer_width", bufferWidth)
                categories?.let { put("categories", it) }
                put("page", page)
                put("limit", limit)
            }

            val response = apiClient.post<Map<String, Any>>(
                "/geographic/route",
                requestBody
            )

            @Suppress("UNCHECKED_CAST")
            val signalsData = response["data"] as List<Map<String, Any>>
            signalsData.map { SignalWithDistance.fromMap(it) }
        } catch (e: Exception) {
            throw ApiException("경로 검색에 실패했습니다: ${e.message}")
        }
    }

    /**
     * 지역 통계
     */
    suspend fun getLocationStatistics(
        latitude: Double,
        longitude: Double,
        radius: Double = 5000.0
    ): GeographicStatistics = withContext(Dispatchers.IO) {
        try {
            val queryParams = mapOf(
                "lat" to latitude.toString(),
                "lon" to longitude.toString(),
                "radius" to radius.toString()
            )

            val response = apiClient.get<Map<String, Any>>(
                "/geographic/statistics",
                queryParams
            )

            @Suppress("UNCHECKED_CAST")
            val statisticsData = response["data"] as Map<String, Any>
            GeographicStatistics.fromMap(statisticsData)
        } catch (e: Exception) {
            throw ApiException("지역 통계 조회에 실패했습니다: ${e.message}")
        }
    }

    /**
     * POI 분석
     */
    suspend fun getPoiAnalysis(
        latitude: Double,
        longitude: Double,
        radius: Double = 2000.0
    ): PoiAnalysisResult = withContext(Dispatchers.IO) {
        try {
            val queryParams = mapOf(
                "lat" to latitude.toString(),
                "lon" to longitude.toString(),
                "radius" to radius.toString()
            )

            val response = apiClient.get<Map<String, Any>>(
                "/geographic/poi-analysis",
                queryParams
            )

            @Suppress("UNCHECKED_CAST")
            val analysisData = response["data"] as Map<String, Any>
            PoiAnalysisResult.fromMap(analysisData)
        } catch (e: Exception) {
            throw ApiException("POI 분석에 실패했습니다: ${e.message}")
        }
    }
}

/**
 * 고급 검색 결과
 */
data class GeographicSearchResult(
    val signals: List<SignalWithDistance>,
    val clusters: List<SignalCluster>? = null,
    val densityMap: List<DensityPoint>? = null,
    val statistics: GeographicStatistics
) {
    companion object {
        fun fromMap(map: Map<String, Any>): GeographicSearchResult {
            @Suppress("UNCHECKED_CAST")
            return GeographicSearchResult(
                signals = (map["signals"] as? List<Map<String, Any>>)
                    ?.map { SignalWithDistance.fromMap(it) } ?: emptyList(),
                clusters = (map["clusters"] as? List<Map<String, Any>>)
                    ?.map { SignalCluster.fromMap(it) },
                densityMap = (map["density_map"] as? List<Map<String, Any>>)
                    ?.map { DensityPoint.fromMap(it) },
                statistics = GeographicStatistics.fromMap(map["statistics"] as Map<String, Any>)
            )
        }
    }
}

/**
 * 시그널 클러스터
 */
data class SignalCluster(
    val id: String,
    val centerLat: Double,
    val centerLon: Double,
    val radius: Double,
    val signalCount: Int,
    val signals: List<SignalWithDistance>? = null,
    val avgDistance: Double
) {
    companion object {
        fun fromMap(map: Map<String, Any>): SignalCluster {
            @Suppress("UNCHECKED_CAST")
            return SignalCluster(
                id = map["id"] as String,
                centerLat = (map["center_lat"] as Number).toDouble(),
                centerLon = (map["center_lon"] as Number).toDouble(),
                radius = (map["radius"] as Number).toDouble(),
                signalCount = map["signal_count"] as Int,
                signals = (map["signals"] as? List<Map<String, Any>>)
                    ?.map { SignalWithDistance.fromMap(it) },
                avgDistance = (map["avg_distance"] as Number).toDouble()
            )
        }
    }
}

/**
 * 밀도 포인트
 */
data class DensityPoint(
    val latitude: Double,
    val longitude: Double,
    val density: Int,
    val weight: Double
) {
    companion object {
        fun fromMap(map: Map<String, Any>): DensityPoint {
            return DensityPoint(
                latitude = (map["latitude"] as Number).toDouble(),
                longitude = (map["longitude"] as Number).toDouble(),
                density = map["density"] as Int,
                weight = (map["weight"] as Number).toDouble()
            )
        }
    }
}

/**
 * 지리적 통계
 */
data class GeographicStatistics(
    val totalSignals: Int,
    val averageDistance: Double,
    val categories: Map<String, Int>,
    val timeDistribution: Map<String, Int>,
    val radialDistribution: List<RadialBucket>
) {
    companion object {
        fun fromMap(map: Map<String, Any>): GeographicStatistics {
            @Suppress("UNCHECKED_CAST")
            return GeographicStatistics(
                totalSignals = map["total_signals"] as Int,
                averageDistance = (map["average_distance"] as Number).toDouble(),
                categories = map["categories"] as Map<String, Int>,
                timeDistribution = map["time_distribution"] as Map<String, Int>,
                radialDistribution = (map["radial_distribution"] as List<Map<String, Any>>)
                    .map { RadialBucket.fromMap(it) }
            )
        }
    }
}

/**
 * 반경별 분포
 */
data class RadialBucket(
    val minDistance: Double,
    val maxDistance: Double,
    val count: Int
) {
    companion object {
        fun fromMap(map: Map<String, Any>): RadialBucket {
            return RadialBucket(
                minDistance = (map["min_distance"] as Number).toDouble(),
                maxDistance = (map["max_distance"] as Number).toDouble(),
                count = map["count"] as Int
            )
        }
    }
}

/**
 * 경로 검색 파라미터
 */
data class RouteSearchParams(
    val startLat: Double,
    val startLon: Double,
    val endLat: Double,
    val endLon: Double,
    val bufferWidth: Double = 1000.0
) {
    fun toMap(): Map<String, Any> {
        return mapOf(
            "start_lat" to startLat,
            "start_lon" to startLon,
            "end_lat" to endLat,
            "end_lon" to endLon,
            "buffer_width" to bufferWidth
        )
    }
}

/**
 * 시간 필터 파라미터
 */
data class TimeFilterParams(
    val startTime: String? = null,
    val endTime: String? = null,
    val timeSlots: List<String>? = null,
    val weekdays: List<Int>? = null
) {
    fun toMap(): Map<String, Any> {
        return buildMap {
            startTime?.let { put("start_time", it) }
            endTime?.let { put("end_time", it) }
            timeSlots?.let { put("time_slots", it) }
            weekdays?.let { put("weekdays", it) }
        }
    }
}

/**
 * 클러스터링 파라미터
 */
data class ClusteringParams(
    val enabled: Boolean = true,
    val distance: Double = 500.0,
    val minPoints: Int = 2
) {
    fun toMap(): Map<String, Any> {
        return mapOf(
            "enabled" to enabled,
            "distance" to distance,
            "min_points" to minPoints
        )
    }
}

/**
 * 핫스팟 정보
 */
data class Hotspot(
    val id: String,
    val name: String,
    val latitude: Double,
    val longitude: Double,
    val signalCount: Int,
    val category: String,
    val score: Double
) {
    companion object {
        fun fromMap(map: Map<String, Any>): Hotspot {
            return Hotspot(
                id = map["id"] as String,
                name = map["name"] as String,
                latitude = (map["latitude"] as Number).toDouble(),
                longitude = (map["longitude"] as Number).toDouble(),
                signalCount = map["signal_count"] as Int,
                category = map["category"] as String,
                score = (map["score"] as Number).toDouble()
            )
        }
    }
}

/**
 * POI 분석 결과
 */
data class PoiAnalysisResult(
    val totalSignals: Int,
    val clusters: List<SignalCluster>,
    val densityPoints: List<DensityPoint>,
    val categoryDistribution: Map<String, Int>,
    val timeDistribution: Map<String, Int>,
    val averageDistance: Double,
    val analysisCenter: Map<String, Double>,
    val analysisRadius: Double
) {
    companion object {
        fun fromMap(map: Map<String, Any>): PoiAnalysisResult {
            @Suppress("UNCHECKED_CAST")
            return PoiAnalysisResult(
                totalSignals = map["total_signals"] as Int,
                clusters = (map["clusters"] as List<Map<String, Any>>)
                    .map { SignalCluster.fromMap(it) },
                densityPoints = (map["density_points"] as List<Map<String, Any>>)
                    .map { DensityPoint.fromMap(it) },
                categoryDistribution = map["category_distribution"] as Map<String, Int>,
                timeDistribution = map["time_distribution"] as Map<String, Int>,
                averageDistance = (map["average_distance"] as Number).toDouble(),
                analysisCenter = map["analysis_center"] as Map<String, Double>,
                analysisRadius = (map["analysis_radius"] as Number).toDouble()
            )
        }
    }
}

File: ios/lib/features/signal/data/services/enhanced_geographic_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/signal_model.dart';
import '../../../../core/network/api_client.dart';
import '../../../../core/network/api_exception.dart';

/// PostGIS 기반 고급 지리적 검색 서비스
class EnhancedGeographicService {
  final ApiClient _apiClient;

  EnhancedGeographicService(this._apiClient);

  /// 고급 지리적 검색
  Future<GeographicSearchResult> advancedSearch({
    required double latitude,
    required double longitude,
    double? radius,
    double? minRadius,
    double? maxRadius,
    List<String>? categories,
    String searchType = 'radius',
    List<List<double>>? polygon,
    RouteSearchParams? route,
    TimeFilterParams? timeFilter,
    bool density = false,
    ClusteringParams? clustering,
    int page = 1,
    int limit = 20,
  }) async {
    try {
      final requestBody = {
        'latitude': latitude,
        'longitude': longitude,
        'radius': radius ?? 5000,
        if (minRadius != null) 'min_radius': minRadius,
        if (maxRadius != null) 'max_radius': maxRadius,
        if (categories != null && categories.isNotEmpty) 'categories': categories,
        'search_type': searchType,
        if (polygon != null) 'polygon': polygon,
        if (route != null) 'route': route.toJson(),
        if (timeFilter != null) 'time_filter': timeFilter.toJson(),
        'density': density,
        if (clustering != null) 'clustering': clustering.toJson(),
        'page': page,
        'limit': limit,
      };

      final response = await _apiClient.post(
        '/geographic/search',
        requestBody,
      );

      return GeographicSearchResult.fromJson(response);
    } catch (e) {
      throw ApiException('고급 검색에 실패했습니다: $e');
    }
  }

  /// 밀도 맵 조회
  Future<List<DensityPoint>> getDensityMap({
    required double latitude,
    required double longitude,
    double radius = 5000,
    double gridSize = 0.01,
  }) async {
    try {
      final response = await _apiClient.get(
        '/geographic/density',
        queryParameters: {
          'lat': latitude.toString(),
          'lon': longitude.toString(),
          'radius': radius.toString(),
          'grid_size': gridSize.toString(),
        },
      );

      final List<dynamic> densityData = response['data'] as List<dynamic>;
      return densityData
          .map((item) => DensityPoint.fromJson(item as Map<String, dynamic>))
          .toList();
    } catch (e) {
      throw ApiException('밀도 맵 조회에 실패했습니다: $e');
    }
  }

  /// 핫스팟 분석
  Future<List<Hotspot>> getHotspots({
    required double latitude,
    required double longitude,
    double radius = 10000,
    int minSignals = 5,
  }) async {
    try {
      final response = await _apiClient.get(
        '/geographic/hotspots',
        queryParameters: {
          'lat': latitude.toString(),
          'lon': longitude.toString(),
          'radius': radius.toString(),
          'min_signals': minSignals.toString(),
        },
      );

      final List<dynamic> hotspotsData = response['data'] as List<dynamic>;
      return hotspotsData
          .map((item) => Hotspot.fromJson(item as Map<String, dynamic>))
          .toList();
    } catch (e) {
      throw ApiException('핫스팟 분석에 실패했습니다: $e');
    }
  }

  /// 다각형 영역 검색
  Future<List<SignalWithDistance>> searchInPolygon({
    required List<List<double>> polygon,
    List<String>? categories,
    int page = 1,
    int limit = 20,
  }) async {
    try {
      final requestBody = {
        'polygon': polygon,
        if (categories != null) 'categories': categories,
        'page': page,
        'limit': limit,
      };

      final response = await _apiClient.post(
        '/geographic/polygon',
        requestBody,
      );

      final List<dynamic> signalsData = response['data'] as List<dynamic>;
      return signalsData
          .map((item) => SignalWithDistance.fromJson(item as Map<String, dynamic>))
          .toList();
    } catch (e) {
      throw ApiException('다각형 검색에 실패했습니다: $e');
    }
  }

  /// 경로 기반 검색
  Future<List<SignalWithDistance>> searchAlongRoute({
    required double startLat,
    required double startLon,
    required double endLat,
    required double endLon,
    double bufferWidth = 1000,
    List<String>? categories,
    int page = 1,
    int limit = 20,
  }) async {
    try {
      final requestBody = {
        'start_lat': startLat,
        'start_lon': startLon,
        'end_lat': endLat,
        'end_lon': endLon,
        'buffer_width': bufferWidth,
        if (categories != null) 'categories': categories,
        'page': page,
        'limit': limit,
      };

      final response = await _apiClient.post(
        '/geographic/route',
        requestBody,
      );

      final List<dynamic> signalsData = response['data'] as List<dynamic>;
      return signalsData
          .map((item) => SignalWithDistance.fromJson(item as Map<String, dynamic>))
          .toList();
    } catch (e) {
      throw ApiException('경로 검색에 실패했습니다: $e');
    }
  }

  /// 지역 통계
  Future<GeographicStatistics> getLocationStatistics({
    required double latitude,
    required double longitude,
    double radius = 5000,
  }) async {
    try {
      final response = await _apiClient.get(
        '/geographic/statistics',
        queryParameters: {
          'lat': latitude.toString(),
          'lon': longitude.toString(),
          'radius': radius.toString(),
        },
      );

      return GeographicStatistics.fromJson(response['data'] as Map<String, dynamic>);
    } catch (e) {
      throw ApiException('지역 통계 조회에 실패했습니다: $e');
    }
  }

  /// POI 분석
  Future<PoiAnalysisResult> getPoiAnalysis({
    required double latitude,
    required double longitude,
    double radius = 2000,
  }) async {
    try {
      final response = await _apiClient.get(
        '/geographic/poi-analysis',
        queryParameters: {
          'lat': latitude.toString(),
          'lon': longitude.toString(),
          'radius': radius.toString(),
        },
      );

      return PoiAnalysisResult.fromJson(response['data'] as Map<String, dynamic>);
    } catch (e) {
      throw ApiException('POI 분석에 실패했습니다: $e');
    }
  }
}

/// 고급 검색 결과
class GeographicSearchResult {
  final List<SignalWithDistance> signals;
  final List<SignalCluster>? clusters;
  final List<DensityPoint>? densityMap;
  final GeographicStatistics statistics;

  GeographicSearchResult({
    required this.signals,
    this.clusters,
    this.densityMap,
    required this.statistics,
  });

  factory GeographicSearchResult.fromJson(Map<String, dynamic> json) {
    return GeographicSearchResult(
      signals: (json['signals'] as List<dynamic>?)
          ?.map((item) => SignalWithDistance.fromJson(item as Map<String, dynamic>))
          .toList() ?? [],
      clusters: (json['clusters'] as List<dynamic>?)
          ?.map((item) => SignalCluster.fromJson(item as Map<String, dynamic>))
          .toList(),
      densityMap: (json['density_map'] as List<dynamic>?)
          ?.map((item) => DensityPoint.fromJson(item as Map<String, dynamic>))
          .toList(),
      statistics: GeographicStatistics.fromJson(json['statistics'] as Map<String, dynamic>),
    );
  }
}

/// 시그널 클러스터
class SignalCluster {
  final String id;
  final double centerLat;
  final double centerLon;
  final double radius;
  final int signalCount;
  final List<SignalWithDistance>? signals;
  final double avgDistance;

  SignalCluster({
    required this.id,
    required this.centerLat,
    required this.centerLon,
    required this.radius,
    required this.signalCount,
    this.signals,
    required this.avgDistance,
  });

  factory SignalCluster.fromJson(Map<String, dynamic> json) {
    return SignalCluster(
      id: json['id'] as String,
      centerLat: (json['center_lat'] as num).toDouble(),
      centerLon: (json['center_lon'] as num).toDouble(),
      radius: (json['radius'] as num).toDouble(),
      signalCount: json['signal_count'] as int,
      signals: (json['signals'] as List<dynamic>?)
          ?.map((item) => SignalWithDistance.fromJson(item as Map<String, dynamic>))
          .toList(),
      avgDistance: (json['avg_distance'] as num).toDouble(),
    );
  }
}

/// 밀도 포인트
class DensityPoint {
  final double latitude;
  final double longitude;
  final int density;
  final double weight;

  DensityPoint({
    required this.latitude,
    required this.longitude,
    required this.density,
    required this.weight,
  });

  factory DensityPoint.fromJson(Map<String, dynamic> json) {
    return DensityPoint(
      latitude: (json['latitude'] as num).toDouble(),
      longitude: (json['longitude'] as num).toDouble(),
      density: json['density'] as int,
      weight: (json['weight'] as num).toDouble(),
    );
  }
}

/// 지리적 통계
class GeographicStatistics {
  final int totalSignals;
  final double averageDistance;
  final Map<String, int> categories;
  final Map<String, int> timeDistribution;
  final List<RadialBucket> radialDistribution;

  GeographicStatistics({
    required this.totalSignals,
    required this.averageDistance,
    required this.categories,
    required this.timeDistribution,
    required this.radialDistribution,
  });

  factory GeographicStatistics.fromJson(Map<String, dynamic> json) {
    return GeographicStatistics(
      totalSignals: json['total_signals'] as int,
      averageDistance: (json['average_distance'] as num).toDouble(),
      categories: Map<String, int>.from(json['categories'] as Map),
      timeDistribution: Map<String, int>.from(json['time_distribution'] as Map),
      radialDistribution: (json['radial_distribution'] as List<dynamic>)
          .map((item) => RadialBucket.fromJson(item as Map<String, dynamic>))
          .toList(),
    );
  }
}

/// 반경별 분포
class RadialBucket {
  final double minDistance;
  final double maxDistance;
  final int count;

  RadialBucket({
    required this.minDistance,
    required this.maxDistance,
    required this.count,
  });

  factory RadialBucket.fromJson(Map<String, dynamic> json) {
    return RadialBucket(
      minDistance: (json['min_distance'] as num).toDouble(),
      maxDistance: (json['max_distance'] as num).toDouble(),
      count: json['count'] as int,
    );
  }
}

/// 경로 검색 파라미터
class RouteSearchParams {
  final double startLat;
  final double startLon;
  final double endLat;
  final double endLon;
  final double bufferWidth;

  RouteSearchParams({
    required this.startLat,
    required this.startLon,
    required this.endLat,
    required this.endLon,
    this.bufferWidth = 1000,
  });

  Map<String, dynamic> toJson() {
    return {
      'start_lat': startLat,
      'start_lon': startLon,
      'end_lat': endLat,
      'end_lon': endLon,
      'buffer_width': bufferWidth,
    };
  }
}

/// 시간 필터 파라미터
class TimeFilterParams {
  final DateTime? startTime;
  final DateTime? endTime;
  final List<String>? timeSlots;
  final List<int>? weekdays;

  TimeFilterParams({
    this.startTime,
    this.endTime,
    this.timeSlots,
    this.weekdays,
  });

  Map<String, dynamic> toJson() {
    return {
      if (startTime != null) 'start_time': startTime!.toIso8601String(),
      if (endTime != null) 'end_time': endTime!.toIso8601String(),
      if (timeSlots != null) 'time_slots': timeSlots,
      if (weekdays != null) 'weekdays': weekdays,
    };
  }
}

/// 클러스터링 파라미터
class ClusteringParams {
  final bool enabled;
  final double distance;
  final int minPoints;

  ClusteringParams({
    this.enabled = true,
    this.distance = 500,
    this.minPoints = 2,
  });

  Map<String, dynamic> toJson() {
    return {
      'enabled': enabled,
      'distance': distance,
      'min_points': minPoints,
    };
  }
}

/// 핫스팟 정보
class Hotspot {
  final String id;
  final String name;
  final double latitude;
  final double longitude;
  final int signalCount;
  final String category;
  final double score;

  Hotspot({
    required this.id,
    required this.name,
    required this.latitude,
    required this.longitude,
    required this.signalCount,
    required this.category,
    required this.score,
  });

  factory Hotspot.fromJson(Map<String, dynamic> json) {
    return Hotspot(
      id: json['id'] as String,
      name: json['name'] as String,
      latitude: (json['latitude'] as num).toDouble(),
      longitude: (json['longitude'] as num).toDouble(),
      signalCount: json['signal_count'] as int,
      category: json['category'] as String,
      score: (json['score'] as num).toDouble(),
    );
  }
}

/// POI 분석 결과
class PoiAnalysisResult {
  final int totalSignals;
  final List<SignalCluster> clusters;
  final List<DensityPoint> densityPoints;
  final Map<String, int> categoryDistribution;
  final Map<String, int> timeDistribution;
  final double averageDistance;
  final Map<String, double> analysisCenter;
  final double analysisRadius;

  PoiAnalysisResult({
    required this.totalSignals,
    required this.clusters,
    required this.densityPoints,
    required this.categoryDistribution,
    required this.timeDistribution,
    required this.averageDistance,
    required this.analysisCenter,
    required this.analysisRadius,
  });

  factory PoiAnalysisResult.fromJson(Map<String, dynamic> json) {
    return PoiAnalysisResult(
      totalSignals: json['total_signals'] as int,
      clusters: (json['clusters'] as List<dynamic>)
          .map((item) => SignalCluster.fromJson(item as Map<String, dynamic>))
          .toList(),
      densityPoints: (json['density_points'] as List<dynamic>)
          .map((item) => DensityPoint.fromJson(item as Map<String, dynamic>))
          .toList(),
      categoryDistribution: Map<String, int>.from(json['category_distribution'] as Map),
      timeDistribution: Map<String, int>.from(json['time_distribution'] as Map),
      averageDistance: (json['average_distance'] as num).toDouble(),
      analysisCenter: Map<String, double>.from(json['analysis_center'] as Map),
      analysisRadius: (json['analysis_radius'] as num).toDouble(),
    );
  }
}